### DISCUSSION TITLE
RNTL v13 & v14 roadmap

### DISCUSSION THREAD

**Discussion by mdjastrzebski** [callstack/react-native-testing-library#1698] [General]
I wanted to share some details about future plans for the React Native Testing Library, especially in the context of the React Test Renderer (RTR) being deprecated. We plan to replace RTR with a new test renderer, for which I’ve already built a prototype that validated this approach.

Here is the RNTL major release roadmap:

-  v12 (current version)
  - Supports React 16, 17, & 18
  - Might work with React 19 (not a goal)
- v13 (Q4 2024)
  - Supports React 18 & 19 only, drops support for 16 & 17
  - Uses React Test Renderer
  - Drops deprecated features (`*ByA11State`, `*ByA11Queries`, etc.)
  - Relatively smooth migration
- v14 (H1 2025)
  - Supports React 19+
  - Uses new test renderer
  - Changes around Fire Event API event handling that might break some code
 
There is no urgency in replacing RTR as it’s still present in React 19, although with some deprecation warnings.

The new test renderer will operate only on host components. In general, this is the desired state, as we already limit RNTL to this. One area where this will impact users is the ability to directly call event handlers on composite components using the Fire Event API. Current RNTL allows this for legacy test reasons, but it will no longer be possible with the new renderer. We will provide an easy migration to v14 for common events like `press` though.

**Comment by mdjastrzebski** [comment]
CC: @thymikee @pierrezimmermannbam @MattAgn @AugustinLF

**Comment by pierrezimmermannbam** [comment]
This sounds like a good plan! 

Regarding V13, you already did pretty much everything right? Besides reviewing and merging #1607 and waiting for RN 0.77 there is nothing else to do, correct?

**Reply by mdjastrzebski** [reply]
I've moved RNTL v13 into beta. I will try to get some feedback from our users, if v13 works fine. Pierre, if you can test it on some of the project you are working on, that would be very helpful. 

From the breaking changes perspective, I think v13 is complete. There seems to be no more legacy staff to drop with this release.

One thing I've considered and kind of rejected was the way OG User Event handles focus, e.g. when you use [`type()`](https://testing-library.com/docs/user-event/utility#type) it will focus on the `<input>`/`<textarea />` element, but it will not trigger `blur` event automatically after the edit. That would only be triggered when some other interaction, e.g. another `type()` or `click()` or [`keyboard("[Tab]")`](https://stackoverflow.com/questions/78384586/how-to-blur-an-input-field-using-testing-library-userevent) would remove the focus from that element. That: 1. seems rather pretty counterintuitive, and 2. could break some tests using `type()` which rely not on `changeText` but rather on `endEditing` (which in RN is basically the same thing as `blur` except for containing the final text).

This subject came to my head in connection with #1677. I still think that some milder version of the above behavior might be beneficial, e.g. we would use "native state" to keep track of focused element. This way, if user passed `skipBlur` to `type()` them it would behave as in the OG User Event.

In general I think such "focus" management would be a nice feature. That being said:
1. There is some research to be done for touch-focus vs keyboard-focus (afair iPads with physical keyboard, probably as well as iPhones,  seem to maintain some sense of element focus).
2. The other part would be supporting `focus`/`blur()` for refs, so that when component code imperatively calls `ref.focus()` / `ref.blur()` we could detect that. That aspect is currently broken due to the way RN mocks components, preventing ref-forwarding, i.e. currently if you request a ref to e.g. `TextInput` you will get class instance of  composite`TextInput` mock instead. What should happen is that RN composite `TextInput` mock should forward ref to host `TextInput` which would trigger the call to `createNodeMock()`. This point probably needs a PR to RN core to fix the mock behaviour by adding ref-forwarding (not a huge change).

If we could both of these, then we would have a really nice focus management. That could happen with v13, but probably would block it for couple of weeks, or we could roll it inside v13 behind feature flag (e.g., `configure({ focusManagement: true })`), and rolled out with v14.

@pierrezimmermannbam I'm curious what are your thoughts on all of these.

**Reply by pierrezimmermannbam** [reply]
I'm trying it on my project, I have failing tests so far, related to concurrent mode and suspense it seems. I'll investigate and let you know once I have more information

**Reply by pierrezimmermannbam** [reply]
Regarding the focus management it's not really clear to me what tracking focus with native state would enable, do you have some examples?
 
I very much agree that solving the issue with refs would be a very good improvement to RN and it should probably done no matter what we do on focus management

**Reply by mdjastrzebski** [reply]
@pierrezimmermannbam send more info on the failing tests, is it v13 or v14 failing? How does it compare to v12 with `concurrentRoot` option set to true?

**Reply by mdjastrzebski** [reply]
re focus tracking: that would be rather marginal update, and leveling-up how RTL/Jest DOM works. We could maintain WebAPI-like [`activeElement`](https://developer.mozilla.org/en-US/docs/Web/API/Document/activeElement), probably as `screen.activeElement` or similar. That way we could track the current active element, whether change came with default event behavior or from user components calling `ref.focus()` in event handlers / effects. Then we could have reliably implement [`toHaveFocus()`](https://github.com/testing-library/jest-dom?tab=readme-ov-file#tohavefocus) matcher and enable users testing focus state.

**Reply by pierrezimmermannbam** [reply]
I'm testing v13-alpha right now, the issues come from concurrentRoot option, if I disable it then it works fine. There are some subtle differences in how suspense works it seems. I managed to fix most of my issues but now I have flaky tests, I need to investigate more to understand what the problem is but it's not related to RNTL probably. I'll try the beta then

**Reply by mdjastrzebski** [reply]
@pierrezimmermannbam I've released v14-alpha using new test renderer that I've wrote (`universal-test-renderer`). It might be somewhat more difficult to check your tests against it. But I am curious about any feedback. This is not urgent, as v13 is still going to happen earlier.
Edit: currently it requires React 19, which might actually prevent you from testing it, util RN 0.77 comes which should use R19.

**Comment by ntucker** [comment]
Is this supposed to be compatible with react 19 and react-test-renderer@19? I'm using react-native@0.78.0-nightly-20250108-ec72af403, which is the first one that lists react 19 as a peerDep, and I'm getting an error when using renderHook:

`Can't access .root on unmounted test renderer`

This works when using RN@77 and react@18.3 and @testing-library/react-native@13

**Reply by mdjastrzebski** [reply]
@ntucker v13 should work with React 19. We have a PR for testing that RNTL works with React 19 + RN nightly: https://github.com/callstack/react-native-testing-library/pull/1723

Could you provide a minimal reproduction repo so I could investigate your issue?

**Comment by edkimmel** [comment]
For v14, can we loosen the restrictions on text elements when running tests configured as `web`? e.g. using the `jest-expo/web` preset?  Text is valid in span/div there, and libraries such as `@expo/html-elements` and `react-strict-dom` do so while being compatible with native.

**Reply by mdjastrzebski** [reply]
The RNTL codebase is not just about queries; it also contains a significant amount of logic that simulates iOS and Android runtime behaviors. When it comes to testing React Native Web apps, using the regular React Testing Library (RTL) is a better fit. It provides a much more realistic browser execution environment. Attempting to add React Native Web simulation to RNTL would result in a lower-quality experience compared to what RTL already offers.

The main downside of this approach is the need for test code duplication. Ultimately, the decision to test web separately or to rely solely on base iOS/Android testing is up to the users. They need to weigh the cost and benefits of such a decision.

One potential solution could be a wrapper library that bridges RNTL and RTL queries into a single API. I reached out to various RN community circles, including Expo, to gauge interest in such a library, but there was little interest.

**Comment by rafaelpadovanirain** [comment]
A big amount of tests started failing in our react native project after the upgrade. Tried `v14.0.0-alpha.1` and even more tests started failing. The same project works perfectly fine with react `v18.3.1`.

**Reply by Noitham** [reply]
We're in the same situation here.
Did you find any solution @rafaelpadovanirain ?

**Reply by mdjastrzebski** [reply]
@rafaelpadovanirain, @Noitham  RNTL 12 & 13 are pretty similar. The main difference is the default value for `concurrentRoot`, try setting it globally to `false` using `configure({ concurrentRoot: false })` in your `jest-setup.ts`. Be aware that with React 19 this setting is mandatory, so you will have to review and fix your tests. 

DO NOT USE RNTL v14 ALPHA, this will definitely not work and it's not supported.

BTW could you post more info about errors you are getting?

See also:
- [this discussion](https://github.com/callstack/react-native-testing-library/discussions/1800)
- [v13 Migration Guide](https://callstack.github.io/react-native-testing-library/docs/migration/v13)

**Reply by mdjastrzebski** [reply]
@rafaelpadovanirain, @Noitham  I've just released [RNTL v13.3 (beta)](https://github.com/callstack/react-native-testing-library/releases/tag/v13.3.0-beta.0) with React Suspense support. 🎉 It adds renderAsync, fireEventAsync, screen.rerenderAsync, etc variants that should work correctly with React Suspense.

Could you test it on your React Suspense code and report back if it works correctly?
